### PR TITLE
UI and API changes to support named flags

### PR DESCRIPTION
**Pull request by kenli-rubrik**
## Description

This pull request adds support for a "name" field in flag objects. It makes the necessary backend and frontend changes.

Backend API changes:
- Add a unique "name" column to the flags table.
- Update flag API to return the name as part of the flag data.
- Update the flag creation API to require a flag name.
- Add a query API (`/query` and `/query/batch`) for getting a flag by name.

UI Changes:
- Add a field for entering flag name when creating flags
- On the front page, show flag names instead of flag descriptions. Also sort by name instead of by ID.
screen shot 2018-06-11 at 10 32 13 am
- On the flag settings page, show the flag name (but make it non-editable).
screen shot 2018-06-11 at 10 19 05 am

## Motivation and Context

We want developers to be able to reference flags by name rather than by ID. It's difficult to maintain flag code if we need to hardcode the ID everywhere, especially when you have projects across multiple languages using the same flags. IDs are granted by the API (auto-increment) so syncing the flags across environments (dev, prod, staging, etc) is also difficult. Having a name field + query API will allow us to query for the ID by name, and then use the ID for other parts of the API (i.e evaluation).

## How Has This Been Tested?

I tested locally by running the API and front-end servers separately. I also built a docker image and tested running as a container.

I also added unit tests for the REST API.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
  (The API documentation is auto-generated from swagger spec. I did not update the README.)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: browser/flagr-ui/src/components/Flag.vue [hunk 1]
```diff
@@ -94,7 +94,7 @@
                 <div slot="header" class="el-card-header">
                   <div class="flex-row">
                     <div class="flex-row-left">
-                      <h2>Flag</h2>
+                      <h2>Flag: {{flag.name}}</h2>
```
Comment (diwu1989) [inline browser/flagr-ui/src/components/Flag.vue]: space between braces
```
{{ flag.name }}
```

File: browser/flagr-ui/src/components/Flag.vue [hunk 2]
```diff
@@ -145,6 +146,13 @@
                       <span size="small">Data Records</span>
                     </el-col>
                   </el-row>
+                  <el-row>
+                    <el-input
```
Comment (diwu1989) [inline browser/flagr-ui/src/components/Flag.vue]: we should try to keep this looking very similar to the existing layout
```
<el-row :gutter="10 class="flag-content">
  <el-col :span="17">
    ...
```
let's add gutter, class to the row, and add col with the correct span

image

File: browser/flagr-ui/src/components/Flag.vue [hunk 3]
```diff
@@ -817,5 +825,6 @@ ol.constraints-inner {
 
 .flag-content{
   margin-top: 8px;
+  margin-bottom: 8px;
```
Comment (diwu1989) [inline browser/flagr-ui/src/components/Flag.vue]: this shouldn't be necessary if `flag-content` was added to the newly added row

File: pkg/handler/query.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+// NewQueryAPI creates a new NewQueryAPI instance
+func NewQueryAPI() QueryAPI {
+	return &queryAPI{}
+}
+
+type queryAPI struct{}
+
+func (qAPI *queryAPI) GetFlagByName(params query.GetFlagByNameParams) middleware.Responder {
+
```
Comment (diwu1989) [inline pkg/handler/query.go]: blank line

File: swagger/query.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+      minLength: 1
+  responses:
+    200:
+      description: returns the flag
+      schema:
+        $ref: "#/definitions/flag"
+    default:
+      description: generic error response
+      schema:
+        $ref: "#/definitions/error"
```
Comment (diwu1989) [inline swagger/query.yaml]: newline at end of file

File: swagger/query_batch.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    200:
+      description: returns list of flags with the given names. Omits flags who's names weren't found.
+      schema:
+        type: array
+        items:
+          $ref: "#/definitions/flag"
+    default:
+      description: generic error response
+      schema:
+        $ref: "#/definitions/error"
```
Comment (diwu1989) [inline swagger/query_batch.yaml]: newline at end of file

**Review by diwu1989 — CHANGES_REQUESTED** [review]
uuid vendor code is included but not used

File: browser/flagr-ui/src/components/Flag.vue (hunk 2 above)
Comment (kenli-rubrik) [reply browser/flagr-ui/src/components/Flag.vue]: Done. I also moved the data records toggle to be next to description again:

New img:
screen shot 2018-06-11 at 11 32 33 am

**Review by diwu1989 — APPROVED** [review]

**Comment by diwu1989** [comment]
extra blank line in `pkg/handler/query.go`

**Pull request merged**